pymeasure/adapters/prologix.py

```diff
@@ -66,10 +67,14 @@ class PrologixAdapter(SerialAdapter):
 
     def __init__(self, port, address=None, rw_delay=None, serial_timeout=0.5,
                  preprocess_reply=None, **kwargs):
+        kwargs.setdefault('write_termination', "\n")
         super().__init__(port, timeout=serial_timeout,
                          preprocess_reply=preprocess_reply, **kwargs)
         self.address = address
         self.rw_delay = rw_delay
+        if rw_delay is not None:
+            warn("Use instrument.ask with a delay argument instead.",
```

Comment:
I'm wondering, is this communication delay between write and read really more an `Instrument` characteristic than an `Adapter` characteristic? 
E.g. that time could be different if I'm using a serial connection or a LAN connection (for multiprotocol instruments), no?
Intuitively, I'd group that delay with other connection configuration like `timeout`.

Reply:
So, maybe `ask` should be in `Adapter`, after all?

Reply:
It might be part of the adapter (which has a delay) or the instrument itself (which responds late) .
In the end, it does not matter, whether I read and wait longer (timeout value) or wait before I try to read.

Keeping ask in the instrument allows to access the instruments read and write, which is necessary in order to make those two functions the main entry points of communication.

Reply:
We could make a hybrid approach: the adapter has the delay variable, but Ask resides in Instrument and sleeps for the sum of the adapter delay and its own parameter.
Alternatively the Adapter could have a delay method (being pass, sleep or doing whatever) which the Instrument ask will call in between read and write

Reply:
> I don't see any use case for waiting before reading. Because you normally read until a term_char (pyvisa) or until a timeout (current SerialAdapter). If the timeout accomodates for the response delay, everything should be fine. If the device responds earlier, you get your data earlier and the code is more simple.

IIRC, there are some instruments where this was useful because something tripped if you `read` too quickly after a `write`, which crashed communication/the device or somesuch. I don't remember the details, though, it would be in the issue tracker somewhere. In any case, here a longer timeout does not help, and the query delay way needed.

> Nevertheless I'm going to include a wait_till_read(query_delay) in the adapter which waits query_delay + self.query_delay (self == adapter) seconds.

Sounds good! Thoughts: 
The one in `Adapter` will probably be set in the communication configuration. How do you imagine that `Instrument` chooses a `query_delay` to pass to that method`? I.e., do we need `Instrument` to add its own delay value?

The name `wait_till_read(query_delay)` strikes me as suboptimal -- that method has nothing to do with reading, just with waiting. I don't yet have a better counterproposal -- `wait_for(query_delay)` would be good but does not express that `self.query_delay` gets added.

Reply:
My thought was, that we have a (potential) delay for the adapter (specific to that and declared in the adapter, maybe overridden) and a delay for the instrument (which always responds late).
As the instrument does not have to take care of the adapter workings, its query can be hard coded into ask.

As the query enters in the middle of Instrument.ask, it has to allow the subclass to call Super.ask with the query parameter.

Reply:
Waiting in 'Instrument.ask' (adding Adapter.query_delay to the wait time) removes the need for naming a method.

A disadvantage would be, that no Adapter method could use the wait. But that should not be a problem, as there should be no need, because the adapter should offer just read and write.
Right now some adapter has a wait for srq method, but this method could (should?) be moved to Instrument, I think.

Reply:
> and a delay for the instrument (which always responds late).

Yeah, but where do we specify it? ` __init__` arg? class attribute?
I haven't used timing-sensitive instruments in a long time, so I have little intuition for what makes most sense.

Reply:
It's rarely used, therefore I propose to not implement it explicitly, but to make an example in the documentation.
You subclass 'ask':
Def ask(self, command):
    "wait half a second before reading" 
    Return Super(). Ask(command, 0.5)

This is simple and easy, I think, without any arguments and such.

The basic idea is, that write, read, and ask contain all communication peculiarities. Waiting is such a peculiarity, therefore that is exactly the place to implement it.

Reply:
Ah, good approach, didn't think of overriding ask()!

Reply:
I think some documentation will be useful that explains the new "philosophy" that you structured! It's getting a lot of moving parts, but nothing that a concise explanation or a diagram can't handle, I think. :-)